arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+    /// column when committing to the witness.
+    pub selectors: Vec<Vec<bool>>,
+
+    /// While folding, we must keep track of the challenges the verifier would
+    /// have sent in the SNARK, and we must aggregate them.
+    // FIXME: nothing is done yet, and the challenges haven't been decided yet.
+    // See top-level documentation of the interpreter for more information.
+    pub challenges: Vec<BigInt>,
+
+    /// Keep the current executed instruction
+    /// This can be used to identify which gadget the interpreter is currently
+    /// building.
+    pub current_instruction: Instruction,
+
+    /// The sponges will be used to simulate the verifier messages, and will
+    /// also be used to verify the consistency of the computation by hashing the
+    /// public IO.
+    // IMPROVEME: use a list of BigInt? It might be faster as the CPU will
+    // already have in its cache the values, and we can use a flat array
+    pub sponge_e1: [BigInt; POSEIDON_STATE_SIZE],
+    pub sponge_e2: [BigInt; POSEIDON_STATE_SIZE],
+
+    /// List of public inputs, used first to verify the consistency of the
+    /// previous iteration.
+    pub current_iteration: u64,
+
+    /// A previous hash, encoded in 2 chunks of 128 bits.
+    pub previous_hash: [u128; 2],
+
+    /// The coin folding combiner will be used to generate the combinaison of
+    /// folding instances
+    pub r: BigInt,
+
+    /// Temporary registers for elliptic curve points in affine coordinates than
+    /// can be used to save values between instructions.
+    ///
+    /// These temporary registers can be loaded into the state by using the
+    /// function `load_temporary_accumulators`.
+    ///
+    /// The registers can, and must, be cleaned after the gadget is computed.
+    ///
+    /// The values are considered as BigInt, even though we should add some
+    /// type. As we want to apply the KISS method, we tend to avoid adding
+    /// types. We leave this for future work.
+    ///
+    /// Two registers are provided, represented by a tuple for the coordinates
+    /// (x, y).
+    pub temporary_accumulators: ((BigInt, BigInt), (BigInt, BigInt)),
+
+    /// Index of the values to absorb in the sponge
+    pub idx_values_to_absorb: usize,
+    // ----------------
+    /// The witness of the current instance of the circuit.
+    /// The size of the outer vector must be equal to the number of columns in the
+    /// circuit.
+    /// The size of the inner vector must be equal to the number of rows in
+    /// the circuit.
+    ///
+    /// The layout columns/rows is used to avoid rebuilding the witness per
+    /// column when committing to the witness.
+    pub witness: Vec<Vec<BigInt>>,
+
+    // --------------
+    // Inputs
+    /// Initial input
+    pub z0: BigInt,
+
+    /// Current input
+    pub zi: BigInt,
+    // ---------------
+
+    // ---------------
+    // Only used to have type safety and think about the design at the
+    // type-level
```

Comment:
you don't need i, the number of the iteration ?

Reply:
Yes. As it was based on Nova initially, `i` appears. We might want to change it.